x/upgrade/internal/types/storeloader.go

```diff
@@ -0,0 +1,129 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+
+	"github.com/spf13/viper"
+
+	"io/ioutil"
+	"os"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// StoreLoaderWithUpgrade is used to prepare baseapp with a fixed StoreLoader
+// pattern. This is useful in test cases, or with custom upgrade loading logic.
+func StoreLoaderWithUpgrade(storeUpgrades *storetypes.StoreUpgrades) baseapp.StoreLoader {
+	return func(ms sdk.CommitMultiStore) error {
+		// TODO cleanup viper
+		home := viper.GetString(flags.FlagHome)
+		upgradeInfoPath := filepath.Join(home, "upgrade-info.json")
+
+		var lastBlockHeight = ms.LastCommitID().Version
+		upgrades, err := GetUpgradeInfoDataFromFile(upgradeInfoPath)
+
+		// There should not be any error in reading upgrade info from filesystem
+		if err != nil {
+			return err
+		}
+
+		// Check if the current commit version and upgrade height matches
+		if (len(upgrades.StoreUpgrades.Renamed) > 0 || len(upgrades.StoreUpgrades.Deleted) > 0) &&
+			upgrades.Height == lastBlockHeight {
+			return ms.LoadLatestVersionAndUpgrade(storeUpgrades)
+		}
+
+		// Otherwise load default store loader
+		return baseapp.DefaultStoreLoader(ms)
+	}
+}
+
+// UpgradeableStoreLoader can be configured by SetStoreLoader() to check for the
+// existence of a given upgrade file - json encoded StoreUpgrades data.
+//
+// If not file is present, it will perform the default load (no upgrades to store).
+//
+// If the file is present, it will parse the file and execute those upgrades
+// (rename or delete stores), while loading the data. It will also delete the
+// upgrade file upon successful load, so that the upgrade is only applied once,
+// and not re-applied on next restart
+//
+// This is useful for in place migrations when a store key is renamed between
+// two versions of the software.
+func UpgradeableStoreLoader(upgradeInfoPath string) baseapp.StoreLoader {
```

Comment:
This is an implementation of the older technique, right?
The code looks correct to me, but I am a bit confused when and where we would use one or the other.

Reply:
This was to address store changes saved to disk, but as that is deprecated I'll remove the one which takes path